server: pause rds when initializing static listeners

snowp opened the pull request:
*title*: server: pause rds when initializing static listeners

*Description*:
In order to prevent RDS requests from being fired up before we know all
the route configurations we care about, pause RDS while we initialize
all the listeners from static config. This matches what we do during LDS
update.

We pause the RDS updates over ADS immediately, then once we know that all 
the init callbacks have been completed we resume RDS updates. This was 
necessary because the RDS subscriptions happen in the init callbacks. 

*Risk Level*: 
Low: Small bug fix

*Testing*:
Manually verified that it fixes the repro case from #3399

*Docs Changes*:
N/A

*Release Notes*:
N/A

Fixes #3399

snowp commented:
I'd love to add a proper test for this but not sure where best to put it. Open for suggestions

junr03 commented:
@htuch assigning over to you due to discussion in #3399

mattklein123 commented on source/server/configuration_impl.cc:
```diff
@@ -51,6 +52,12 @@ void MainImpl::initialize(const envoy::config::bootstrap::v2::Bootstrap& bootstr
+  // Immediately pause RDS to ensure that we don't send any requests until we've
+  // subscribed to all the RDS resources. The subscriptions happen in the init callbacks,
+  // so we pause RDS until we've completed all the callbacks.
+  cluster_manager_->adsMux().pause(Config::TypeUrl::get().RouteConfiguration);
```
I don't have a great answer, but I wish there was a more general way of solving this problem, as I could see us having issues in the future for other resource types that get added. @htuch any ideas?

htuch replied:
I think it's fine, there are only two other places this comes up: EDS and SDS. We should either fix EDS in this PR or leave some TODOs and an open issue. I don't think we're going to arbitrarily scale the number of resources in xDS, so seems fine to have a handful of special case handling.

htuch commented on source/server/server.cc:
```diff
@@ -388,7 +389,7 @@ RunHelper::RunHelper(Event::Dispatcher& dispatcher, Upstream::ClusterManager& cm
   // this can fire immediately if all clusters have already initialized. Also note that we need
   // to guard against shutdown at two different levels since SIGTERM can come in once the run loop
   // starts.
-  cm.setInitializedCb([this, &init_manager, workers_start_cb]() {
```
I think the scope here is too wide. Won't this prevent listeners warming? I think ideally we pause/resume around just this loop: https://github.com/envoyproxy/envoy/blob/c223232bbff5aadde8ec83a548e18f99d4729914/source/server/configuration_impl.cc#L54

snowp replied:
I tried that at first but it didn't seem to work - while the listeners are created there the RDS subscription isn't started until the initialize callback. Pausing just around the static listener loop mean that we resumed before the subscriptions started.

It won't block the listeners from warming because I'm resuming after all `initialize` callbacks are called, not when initialization has been complete. Doing the resume inside the callback passed to `init_manager` does prevent the listeners from warming, presumably because some init callback is waiting for the routes. 

Happy to consider another approach - I'm not super happy about having the start and resume action be so far from each other.

snowp commented:
Added a test to verify that this fixes it for RDS and that it already works for EDS. Also moved the `pause` call to the ClusterManager's init callback: all the RDS subscriptions happen in the init callbacks, so we can narrow the pause down to just around the init callbacks. 

Turns out this already works for EDS: the subscriptions happen during `ClusterManager` init *before* the ADS mux has been started. Once the `GrpcMux` starts, it fires off a xDS request for all the resources, at which point all the statically defined EDS resources have already been subscribed to.

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -489,5 +493,112 @@ TEST_P(AdsConfigIntegrationTest, EdsClusterWithAdsConfigSource) {
+  - name: dummy_listener
+    address:
+      socket_address:
+        address: 127.0.0.1
```
I'd suggest taking the existing bootstrap built by the config utility helper and just augmenting the static resources in `ConfigHelper::addConfigModifier`. That will reduce boilerplate here and make the test more robust as we evolve the integration test bootstrap.

htuch commented on test/integration/ads_integration_test.cc:
```diff
@@ -489,5 +493,112 @@ TEST_P(AdsConfigIntegrationTest, EdsClusterWithAdsConfigSource) {
+};
+
+TEST_P(StaticAdsIntegrationTest, XdsBatching) {
+  // Validates that the initial xDS request batches all resources referred to in static config
```
Can you move this to above the `TEST_P`?
